x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java

```diff
@@ -607,16 +608,41 @@ private DeleteRequest createRoleDeleteRequest(final String roleName) {
         return client.prepareDelete(SECURITY_MAIN_ALIAS, getIdForRole(roleName)).request();
     }
 
-    private XContentBuilder createRoleXContentBuilder(RoleDescriptor role) throws IOException {
+    // Package private for testing
+    XContentBuilder createRoleXContentBuilder(RoleDescriptor role) throws IOException {
         assert NativeRealmValidationUtil.validateRoleName(role.getName(), false) == null
             : "Role name was invalid or reserved: " + role.getName();
         assert false == role.hasRestriction() : "restriction is not supported for native roles";
-        return role.toXContent(
-            jsonBuilder(),
-            ToXContent.EMPTY_PARAMS,
-            true,
-            featureService.clusterHasFeature(clusterService.state(), SECURITY_ROLES_METADATA_FLATTENED)
-        );
+
```

Comment:
I went back and forth on this but landed on that the cleanest approach is to handle the special case for the security index defaults and `metadata_flattened` logic here. 

Simply using the `docCreation` is `true` is not enough since it's also used in the [ApiKeyService](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L2028) in a couple of places, to create role descriptors. Since we don't do the same upsert operations on api keys, the default values are not needed and the `metadata_falttened` field is not needed. This could be fixed by providing a couple of extra parameters to `toXContent` or a context object of some sort, that would need to be populated by all callers. In the end this turned in to many booleans, instead of actually doing the logic where we have all the context. 

The downside with this approach is that the role descriptor logic is now spread out.